crates/matrix-sdk/src/room_preview.rs

```diff
@@ -69,19 +69,46 @@ pub struct RoomPreview {
     ///
     /// Set to `None` if the room is unknown to the user.
     pub state: Option<RoomState>,
+
+    /// The `m.room.direct` state of the room, if known.
+    pub is_direct: Option<bool>,
+
+    /// The [`Client`] that can be used to perform actions given the room
+    /// preview.
+    pub(crate) client: Client,
 }
 
 impl RoomPreview {
+    /// Leave a room using its [`RoomPreview`].
+    pub async fn leave(&self) -> Result<(), Error> {
+        match self.state {
+            Some(RoomState::Left) | None => {
+                Err(Error::WrongRoomPreviewState(WrongRoomPreviewState {
+                    expected: "Invited, Knocked or Joined",
+                    got: self.state,
+                }))
+            }
+            Some(_) => self.client.leave_room(&self.room_id).await.map_err(Into::into),
+        }
+    }
```

Comment:
Sorry, I still think that `RoomPreview` ought to be a purely descriptive data record (i.e. no action methods); can the caller use the room id that's always around, and call `leave_room` by themselves? Then, can we get rid of the `Client` stored and address the other comment I had on `from_room_info`?